### DISCUSSION TITLE
Failure to connect to Database.

### DISCUSSION THREAD

**Discussion by Dvalin21** [dani-garcia/vaultwarden#3423] [Q&A] [answered]
I initially installed this before 202303 update and had no problems. Since the update to the latest version 1.28, I've been getting the following error from Vaultwarden container

2023-04-06T03:48:06.418147757Z | stdout | "could not connect to server: Connection refused\n\tIs the server running on host \"vaultwarden-db\" (172.22.0.3) and accepting\n\tTCP/IP connections on port 5432?\n",
-- | -- | --
2023-04-06T03:48:06.418062647Z | stdout | [CAUSE] BadConnection(
2023-04-06T03:48:06.417418116Z | stdout | [2023-04-06 03:48:06.417][vaultwarden][ERROR] Error creating database pool: DieselCon.

And then from VW-DB

2023-04-06T03:20:10.061311328Z | stderr | 2023-04-06 03:20:10.061 UTC [1] LOG:  database system is ready to accept connections
-- | -- | --
2023-04-06T03:20:09.921517835Z | stderr | 2023-04-06 03:20:09.921 UTC [17] LOG:  database system was shut down at 2023-04-06 03:19:35 UTC
2023-04-06T03:20:09.082766057Z | stderr | 2023-04-06 03:20:09.082 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-04-06T03:20:08.532571565Z | stderr | 2023-04-06 03:20:08.532 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2023-04-06T03:20:08.532422279Z | stderr | 2023-04-06 03:20:08.532 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-04-06T03:20:08.474777282Z | stderr | 2023-04-06 03:20:08.474 UTC [1] LOG:  starting PostgreSQL 15.2 (Debian 15.2-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
2023-04-06T03:20:06.051806769Z | stdout |  
2023-04-06T03:20:06.051716946Z | stdout | PostgreSQL Database directory appears to contain a database; Skipping initialization
2023-04-06T03:20:06.051558186Z | stdout |  
(I have a feeling this is is the issue)2023-04-06T03:20:06.040152011Z | stderr | chmod: changing permissions of '/var/run/postgresql': Operation not permitted

And here is a portions of my docker compose to help with the issue:

The storage location permission is set to :rw
Should I set the container for the database to root?

**Comment by Dvalin21** [comment]
Update: Fixed the chmod error, but still get the connection error regardless.

**Reply by stefan0xC** [reply]
How did you fix the permission error? What does the database container logs say now?

**Reply by Dvalin21** [reply]
> How did you fix the permission error? What does the database container logs say now?

Ran both containers as root

**Reply by Dvalin21** [reply]
However this wasn't needed before.

**Comment by Dvalin21** [comment]
Update: I've completely removed Vaultwarden and database and reinstalled...same issue. DieselCon keeps retrying on port 5432.  The only change before all of this started was the update. Can you verify? Thanks in advance.

**Comment by BlackDex** [comment]
I don't know what the previous version was you were running. But if that was 1.27, there is nothing changed between those versions in regards to the database code.

So i think it maybe has to do something with docker or something else in the network maybe?

**Reply by Dvalin21** [reply]
There hasn't been anything that I've changed to either Vaultwarden or VW-DB. No network changes on my end. However I am aware that the db did do an update at the same time Vaultwarden did. Vaultwarden stops because of the error

**Comment by Dvalin21** [comment]
Update: I've been searching and found that others were having the same issues. They were migrating from mysql to post https://vaultwarden.discourse.group/t/unix-socket-for-postgres/1841

**Reply by BlackDex** [reply]
That isn't comparable, since you are not using a socket but an IP connection. So this is not the same.

**Comment by Dvalin21** [comment]
There's nothing that would show other than on the Vaultwarden side that shows an error. There has to be something between the two. I've completely moved them both and completely reconstructed them only to get the same error. Changed the network setting for the container, still the same thing.

**Reply by BlackDex** [reply]
But that isn't how you configured Vaultwarden. And besides that. a socket is within the container of PostgreSQL, and never reachable for Vaultwarden unless  you share that socket via volumes

**Reply by Dvalin21** [reply]
> But that isn't how you configured Vaultwarden. And besides that. a socket is within the container of PostgreSQL, and never reachable for Vaultwarden unless you share that socket via volumes

Here is an example of how I installed vaultwarde (docker compose). With this setup, everything was working until today

version: "3.9"
services:
  db:
    image: postgres
    container_name: Vaultwarden-DB
    hostname: vaultwarden-db
    mem_limit: 512m
    cpu_shares: 768
    security_opt:
      - no-new-privileges:true
    user: 1026:100
    healthcheck:
      test: ["CMD", "pg_isready", "-q", "-d", "vaultwarden", "-U", "vaultwardenuser"]
      timeout: 45s
      interval: 10s
      retries: 10
    volumes:
      - /volume1/docker/vaultwarden/db:/var/lib/postgresql/data:rw
    environment:
      POSTGRES_DB: vaultwarden
      POSTGRES_USER: vaultwardenuser
      POSTGRES_PASSWORD: vaultwardenpass
    restart: on-failure:5

  vaultwarden:
    image: vaultwarden/server:latest
    container_name: Vaultwarden
    hostname: vaultwarden
    mem_limit: 256m
    mem_reservation: 96m
    cpu_shares: 1024
    security_opt:
      - no-new-privileges:true
    user: 1026:100
    ports:
      - 4080:4020
    volumes:
      - /volume1/docker/vaultwarden/data:/data:rw
    environment:
      ROCKET_PORT: 4020
      DATABASE_URL: postgresql://vaultwardenuser:vaultwardenpass@vaultwarden-db:5432/vaultwarden
      ADMIN_TOKEN: MariushostingMariushostingMari13
      DISABLE_ADMIN_TOKEN: false
      DOMAIN: https://vaultwarden.yourname.synology.me
      SMTP_HOST: smtp.gmail.com
      SMTP_FROM: Your-own-gmail-address
      SMTP_PORT: 587
      SMTP_SECURITY: starttls
      SMTP_USERNAME: Your-own-gmail-address
      SMTP_PASSWORD: Your-own-app-password
    restart: on-failure:5
    depends_on:
      db:
        condition: service_started

Just in case I got it from here. https://mariushosting.com/synology-install-vaultwarden-with-portainer/
And yes I am using a Synology/Xpenology NAS.

What is the DieselCon?  Apparently its not the db that is using port 5432, but this function?

**Comment by BlackDex** [comment]
Try to add this to the postgres config.
```yaml
ports:
 - 5432:5432
```

It's as simple as that Vaultwarden just can't connect to the Postgresql database, nothing more nothing less.

**Reply by Dvalin21** [reply]
> Try to add this to the postgres config.
> 
> ```yaml
> ports:
>  - 5432:5432
> ```
> 
> It's as simple as that Vaultwarden just can't connect to the Postgresql database, nothing more nothing less.

Tried that and it would say the port is already in use.

**Reply by BlackDex** [reply]
That would mean an other instance is already running on that port, or something else.

**Reply by Dvalin21** [reply]
> That would mean an other instance is already running on that port, or something else.

I keep seeing something same DieselCon retry when the connection fails. There's no other container on my host thats using that port. I haven't installed anything new that would take the port. This was running on that port before the update.

**Reply by stefan0xC** [reply]
Are you trying to connect to a postgres database that is running on the host?

**Reply by Dvalin21** [reply]
Yes, both containers are on the same host. FYI just ran the following and it came back with nothing using port 5432 netstat -pat | grep LISTEN.

**Reply by Dvalin21** [reply]
Something I forgot before. Before I had completely delete and reinstalled Vaultwarden, from the database I was getting an error that said it couldn't send incomplete 2fa. I never setup the 2fa on the database in the first place. Again, all after the update. Which I know they updated the same time Vault did

**Reply by BlackDex** [reply]
> Yes, both containers are on the same host. FYI just ran the following and it came back with nothing using port 5432 netstat -pat | grep LISTEN.

If nothing is running, then Vaultwarden can't connect.
I would suggest to focus on that, and try my suggestion below.

**Comment by Dvalin21** [comment]
Running the database container as root. All it says now is listening to
port 0000:5432. However Vaultwarden keeps stopping due to the connection
refused. Then it will say retry DieselCon

**Reply by BlackDex** [reply]
Try an other client which is able to connect to PostgreSQL, like DBeaver, and see if you can connect your self.

**Reply by Dvalin21** [reply]
Would MariaDB work?
Also, I changed the port 5432 to 5433 just to see what would happen. Same issue.

**Reply by stefan0xC** [reply]
Make sure that that the `db` container is running (what does `docker ps` say?) and that the hostname in the `DATABASE_URL` matches what you set for the `postgres` service (if you have changed the name from the example). You might also want to either explicitly [link the containers](https://docs.docker.com/compose/networking/#link-containers) or change the hostname inside the `DATABASE_URL` variable from `vaultwarden-db` to just `db`  as the vaultwarden service should be able to connect to the service by that name.

**Reply by Dvalin21** [reply]
> Make sure that that the `db` container is running (what does `docker ps` say?) and that the hostname in the `DATABASE_URL` matches what you set for the `postgres` service (if you have changed the name from the example). You might also want to either explicitly [link the containers](https://docs.docker.com/compose/networking/#link-containers) or change the hostname inside the `DATABASE_URL` variable from `vaultwarden-db` to just `db` as the vaultwarden service should be able to connect to the service by that name.

Just verified everything. One of the hostnames were incorrect, I did fix it and restarted, but still get the refuse connection to the database. However the only difference is that vaultwarden is not shutting down like it did before. I know that when I first installed it, I didnt have to use root. I do however have to run the database as root in order for it to run correctly. I did see there was a change or a commit that for postgres just two weeks ago that maybe causing the issue. 

Also I just noticed this error when looking at the container: curl: (7) Failed to connect to localhost port 4020: Connection refused

**Reply by Dvalin21** [reply]
> Try an other client which is able to connect to PostgreSQL, like DBeaver, and see if you can connect your self.

Do you have a docker-compose or example of one to setup dbeaver with Vaultwarden?

**Comment by Dvalin21** [comment]
@BlackDex So I have no clue how I was at least able to get this to run (kinda), but yes, I was able to connect to Cloudbeaver. So the issue is postgres. Do you have a compose that is setup correctly for vaultwarden? Thanks.

**Comment by Dvalin21** [comment, marked as answer]
So after a few days I've figured out my entire issue.

1. Postgres did an update recently that is causing duplicate port use 5432 (As per @BlackDex switched to Dbeaver database, works great)
2. Also, update firewall rules from the following article https://mariushosting.com/how-to-set-up-synology-firewall-geoip-blocking/
Its a great, except what should be added is port 53 (DNS) and if you are using it Port 587 (Starttls)
3. Make sure when you add these to your firewall rules, that you select protocol "All" for both and you place them at the top of the list. THE ORDER MATTERS.
